src/Renderer/LayeredMaterial.js

```diff
@@ -74,14 +74,46 @@ var moveElementsArraySafe = function moveElementsArraySafe(array,index, howMany,
 };
 /* eslint-enable */
 
+const arrayColorLength = 15;
+const nbSamplersElevation = 1;
+
+function getCountColorLayer() {
+    const maxTexturesUnits = Capabilities.getMaxTextureUnitsCount();
+    const nbSamplersColor = Math.min(maxTexturesUnits - nbSamplersElevation);
+    const sizeColor_01 = nbSamplersColor > arrayColorLength ? arrayColorLength : nbSamplersColor;
+    const sizeColor_02 = nbSamplersColor > arrayColorLength ? arrayColorLength : 0;
+    return { sizeColor_01, sizeColor_02 };
+}
+
+export function getMaxColorLayerCount() {
+    const { sizeColor_01, sizeColor_02 } = getCountColorLayer();
+    return sizeColor_01 + sizeColor_02;
+}
+
 const LayeredMaterial = function LayeredMaterial(options) {
     THREE.RawShaderMaterial.call(this);
 
-    const maxTexturesUnits = Capabilities.getMaxTextureUnitsCount();
-    const nbSamplers = Math.min(maxTexturesUnits - 1, 16 - 1);
```

Comment:
Unfortunately we cannot remove the `16-1` part yet, at least not on linux, not until https://bugzilla.mozilla.org/show_bug.cgi?id=777028 is fixed... :-/

Reply:
Oh forget this, you still have it on `arrayColorLength` I guess.

Reply:
is with Mesa?

Reply:
Actually, I was right: on max, you can have `arrayColorLength * 2` samplers. So bug above applies.

Reply:
no, I don't think so
[because we test if we are in mesa driver](https://github.com/iTowns/itowns/pull/583/files#diff-dd48a35666ea82b037874482f9b61b60R23)

Reply:
Cool!

Reply:
but in fact it doesn't work
I am looking for a way to detect if mesa is present

Reply:
I fixe the firefox / mesa bug
unfortunately, the last test commit with roll loop doens't work in mesa